packages/babel-core/src/transform-ast.js

```diff
@@ -1,21 +1,56 @@
 // @flow
-import * as t from "@babel/types";
-import loadConfig from "./config";
-import runTransform, { type FileResult } from "./transformation";
-
-export default function transformFromAst(
-  ast: Object,
-  code: string,
-  opts: Object,
-): FileResult | null {
-  const config = loadConfig(opts);
-  if (config === null) return null;
-
-  if (ast && ast.type === "Program") {
-    ast = t.file(ast, [], []);
-  } else if (!ast || ast.type !== "File") {
-    throw new Error("Not a valid ast?");
+
+import loadConfig, { type InputOptions } from "./config";
+import {
+  runAsync,
+  type FileResult,
+  type FileResultCallback,
+} from "./transformation";
+
+import transformAstSync from "./transform-ast-sync";
+
+type AstRoot = BabelNodeFile | BabelNodeProgram;
+
+type TransformAst = {
+  (ast: AstRoot, code: string, callback: FileResultCallback): void,
+  (
+    ast: AstRoot,
+    code: string,
+    opts: ?InputOptions,
+    callback: FileResultCallback,
+  ): void,
+
+  // Here for backward-compatibility. Ideally use ".transformSync" if you want
+  // a synchronous API.
+  (ast: AstRoot, code: string, opts: ?InputOptions): FileResult | null,
+};
+
+export default ((function transformFromAst(ast, code, opts, callback) {
```

Comment:
Do we need to emulate an async function? What about not providing this at all? Only `transformFromAstSync` or `transformFromAst` which is sync?

Otherwise lgtm.

Reply:
I think we need to emulate the async function so that we can in the future introduce async plugins without breaking changes.

Reply:
Good point :+1:

Reply:
Yeah exactly. My goal with this PR is just to expose the API so tooling can start using it, even if the plugins don't support async yet. This way we at least have more of a migration path.